pytorch_lightning/strategies/bagua.py

```diff
@@ -0,0 +1,207 @@
+from pytorch_lightning.strategies import DDPStrategy
+from pytorch_lightning.utilities.enums import _StrategyType
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.imports import _BAGUA_AVAILABLE
+from pytorch_lightning.utilities.seed import reset_seed
+
+if _BAGUA_AVAILABLE:
+    import bagua.torch_api as bagua
+    from bagua.torch_api.algorithms import Algorithm
+    from bagua.torch_api.algorithms.q_adam import QAdamOptimizer
+    from bagua.torch_api.communication import allreduce_inplace, barrier, broadcast_object, is_initialized, ReduceOp
+    from bagua.torch_api.data_parallel.distributed import DistributedDataParallel_V1_9_0 as BaguaDistributedDataParallel
+
+
+log = logging.getLogger(__name__)
+
+
+class LightningBaguaModule(_LightningModuleWrapperBase):
+    def __init__(self, pl_module: "pl.LightningModule") -> None:
+        super().__init__(pl_module)
+        # Bagua use `bagua_module_name` to distinguish different modules
+        self._bagua_module_name = pl_module._get_name() + str(id(pl_module))
+
+    def forward(self, *inputs, **kwargs):
+        return super().forward(*inputs, **kwargs)
+
+
+class BaguaStrategy(DDPStrategy):
+
+    distributed_backend = _StrategyType.BAGUA
+
+    def __init__(
+        self,
+        algorithm: str = "gradient_allreduce",
+        gradient_as_bucket_view: bool = True,
+        parallel_devices: Optional[List[torch.device]] = None,
+        cluster_environment: Optional[ClusterEnvironment] = None,
+        checkpoint_io: Optional[CheckpointIO] = None,
+        precision_plugin: Optional[PrecisionPlugin] = None,
+        **kwargs: Union[Any, Dict[str, Any]],
+    ):
+
+        super().__init__(
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+            checkpoint_io=checkpoint_io,
+            precision_plugin=precision_plugin,
+        )
+
+        self._bagua_algorithm = algorithm
+        self._bagua_gradient_as_bucket_view = gradient_as_bucket_view
+        self._bagua_kwargs = kwargs
+
+    def setup_environment(self) -> None:
+        # start the other scripts
+        if not self.cluster_environment.creates_processes_externally:
+            self._call_children_scripts()
+
+        self.setup_distributed()
+
+    def setup_distributed(self):
+        reset_seed()
+
+        # determine which process we are and world size
+        self.set_world_ranks()
+
+        self._init_bagua_distributed()
+
+    def setup(self, trainer: "pl.Trainer") -> None:
+        super().setup(trainer)
+
+        # move the model to the correct device
+        self.model_to_device()
+
+        model = LightningBaguaModule(self.model)
+        self.configure_bagua_ddp(model)
+
+        # start the background communication for async algorithm
+        if self._bagua_algorithm == "async":
+            self.model.bagua_algorithm.resume(self.model)
```

Comment:
I moved this code from pre_dispatch to here (pre_dispatch removed on master). Now, some adjustments need to be discussed here. I recommend we look at super().setup(). There we already call model_to_device(). Instead of implementing `configure_bagua_ddp`, we could just implement `configure_ddp()` maybe? 

Does Bagua want to wrap the model also for inference, or should it only be done for training?

Reply:
Yes, we can implement `configure_ddp()` instead.

Bagua can optionally wrap the model for inference, they have the same behavior.